File: backend/src/main/java/ca/bc/gov/nrs/ilcr/millcontext/dto/WorkingContext.java
package ca.bc.gov.nrs.ilcr.millcontext.dto;

/**
 * The resolved Home working context for a selected (mill, reporting year) pair (Story 1.2, pinned
 * wire contract AD-12 — unchanged from the Story 1.1 pin). Returned by {@code GET
 * /api/v1/mill-context}; the Home page (1.3) confirms the selection with it and the banner (1.4)
 * renders it.
 *
 * <p>Either track status may be null: no {@code THE.ILCR_MILL_REPORT_STATUS} row for the pair (S07)
 * nulls both; a NULL code column nulls just that track (legacy would NPE on a NULL silviculture
 * code — tolerated here, deviation recorded in Story 1.2). Under global Jackson {@code non_null} a
 * null status is omitted from the JSON. {@code millNumber}/{@code millName} may be null (legacy
 * nullable columns — see the 1.1 contract nullability note).
 *
 * <p>Story 1.3 amendment (AD-12): {@code message} carries the SUC-001 confirmation ({@code
 * dataSavedSuccesfullyInfoMsg} → "Data saved successfully") on EVERY 200, mirroring how Schedule
 * 1's save sources its success text server-side (AD-8). Semantic caveat: the endpoint is used both
 * to CONFIRM a Home save (1.3) and to LOAD the banner (1.4); the message is present on every 200
 * but the frontend must only DISPLAY it after an explicit Save — 1.4's banner load ignores it.
 * Backward-compatible because the change is purely ADDITIVE — existing 1.2 consumers just see one
 * extra field they do not read. (Note {@code message} is populated on every 200, so the global
 * {@code non_null} setting never omits it; compatibility does not rest on omission.)
 *
 * @param millId the mill id ({@code THE.MILL.MILL_ID})
 * @param millNumber the mill number as a display string; may be null
 * @param millName the mill name; may be null
 * @param reportYear the selected reporting year
 * @param schedules1To10Status the Schedules 1–10 track status; null when none (S07)
 * @param schedule11Status the Schedule 11 (silviculture) track status; null when none
 * @param millViewable false when the mill's {@code ILCR_MILL_STATUS_XREF} status is not {@code ACT}
 *     (closed mills stay selectable — S06 — but schedule pages block viewing)
 * @param message the SUC-001 confirmation (key + server-resolved verbatim text); Story 1.3 (AC7)
 */
public record WorkingContext(
    long millId,
    String millNumber,
    String millName,
    int reportYear,
    TrackStatus schedules1To10Status,
    TrackStatus schedule11Status,
    boolean millViewable,
    MessageInfo message) {}

File: frontend/src/interfaces/WorkingContext.ts
// Mirrors the backend millcontext WorkingContext DTO (Stories 1.1/1.2 pinned wire contract,
// GET /v1/mill-context), AMENDED by Story 1.3 to carry the SUC-001 `message` on every 200 (AC7).
//
// Jackson `non_null` omits any null field, so millNumber/millName, both track statuses, and message
// are all optional — render defensively. The status fields feed Story 1.4's banner; the full type is
// kept here so 1.4 reuses it unchanged. `message` is present on every 200 but is only DISPLAYED
// after an explicit Save (1.4's banner load must ignore it).

// Success message carried on the 200 response (AD-8): the frontend renders `text` verbatim and never
// hardcodes SUC-* strings. Mirrors interfaces/Schedule1Response.ts MessageInfo.
export interface MessageInfo {
  readonly key: string
  readonly text: string
}

// One workflow track's status (Schedules 1–10 or Schedule 11). All members are nullable/omitted.
export interface TrackStatus {
  readonly code?: string | null
  readonly description?: string | null
  readonly date?: string | null
}

export default interface WorkingContext {
  readonly millId: number
  readonly millNumber?: string | null
  readonly millName?: string | null
  readonly reportYear: number
  readonly schedules1To10Status?: TrackStatus | null
  readonly schedule11Status?: TrackStatus | null
  readonly millViewable: boolean
  readonly message?: MessageInfo | null
}

// The RFC 7807 problem+json body returned on 400 (S04/S05/S08) and other error paths. `messages`
// carries the per-field verbatim texts (both when both fields are missing); `detail` is the joined
// fallback. The frontend renders `messages[].text`, falling back to `detail` — never hardcoded.
export interface ProblemBody {
  readonly detail?: string
  readonly messages?: readonly MessageInfo[]
}
